### PR TITLE
CryptoCreate to work with ECDSA public key aliases

### PR DESCRIPTION
**Pull request by tdermendzhievv**
**Description**:

This PR introduces a logic for crypto create with alias. Covered cases:
cryptoCreate with EC key alias
cryptoCreate with EVM address alias
cryptoCreate with ED key alias

**Related issue(s)**: #3955 #3954 

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 1]
```diff
@@ -119,6 +134,19 @@ public void doStateTransition() {
...
+                aliasManager.maybeLinkEvmAddress(jKey, EntityNum.fromAccountId(created));
+            }
+            if (op.hasKey() && !op.getKey().getECDSASecp256K1().isEmpty()) {
+                aliasManager.link(
+                        ledger.getAccountsLedger().getImmutableRef(created).getAlias(),
+                        EntityNum.fromAccountId(created));
+            }
         } catch (InsufficientFundsException ife) {
             txnCtx.setStatus(INSUFFICIENT_PAYER_BALANCE);
         } catch (Exception e) {
```
Comment (dimitar-dinev) [inline hedera-node/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:152]: I think we'll need to add a catch when we throw `InvalidTransactionException` from an already used alias because currently in these scenarios we come here and `FAIL_INVALID` is set. 

I'm also wondering whether we should check if aliases are in use in the semantic checks 🤔

File: hedera-node/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 2]
```diff
@@ -132,17 +160,122 @@ private HederaAccountCustomizer asCustomizer(CryptoCreateTransactionBody op) {
...
+                        .autoRenewPeriod(THREE_MONTHS_IN_SECONDS)
+                        .expiry(txnCtx.consensusTime().getEpochSecond() + THREE_MONTHS_IN_SECONDS)
+                        .isReceiverSigRequired(false)
+                        .isSmartContract(false)
+                        .alias(op.getAlias());
+            }
+        } else if (!op.getAlias().isEmpty() && op.hasKey()) {
+            if (!aliasManager.lookupIdBy(op.getAlias()).equals(MISSING_NUM)) {
+                throw new InvalidTransactionException(INVALID_ALIAS_KEY);
+            }
```
[the remark this answers is not in the document]
Comment (IvanKavaldzhiev) [reply hedera-node/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:206]: Or `throwIfAliasMissing`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 3]
```diff
@@ -119,8 +134,24 @@ public void doStateTransition() {
...
+                aliasManager.link(op.getAlias(), EntityNum.fromAccountId(created));
+                if (op.getAlias().size() > EVM_ADDRESS_SIZE) {
+                    final var key = asPrimitiveKeyUnchecked(op.getAlias());
+                    JKey jKey = asFcKeyUnchecked(key);
+                    aliasManager.maybeLinkEvmAddress(jKey, EntityNum.fromAccountId(created));
+                }
+            } else {
+                if (op.hasKey() && !op.getKey().getECDSASecp256K1().isEmpty()) {
+                    aliasManager.link(
+                            ledger.getAccountsLedger().getImmutableRef(created).getAlias(),
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:147]: Q: Should this be checking for a valid non null alias before doing the link?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 4]
```diff
@@ -132,17 +163,100 @@ private HederaAccountCustomizer asCustomizer(CryptoCreateTransactionBody op) {
...
+                                .lookupIdBy(
+                                        ByteString.copyFrom(recoveredEvmAddressFromPrimitiveKey))
+                                .equals(MISSING_NUM)) {
+                    throw new InvalidTransactionException(INVALID_ALIAS_KEY);
+                }
+                customizer.alias(ByteString.copyFrom(recoveredEvmAddressFromPrimitiveKey));
+            }
+
+            final JKey key = asFcKeyUnchecked(op.getKey());
+            customizer
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:251]: There's differences and similarities between this customizer set and the other  cases in terms of fields and values.
We should pull out the shared set to be at the beginning and let the if/else set or override the special a cases.

e.g.
- `isDeclinedReward` and `maxAutomaticAssociations ` are only set here
- this is the only case that sets expiry to `expiry`. The other cases set it as `txnCtx.consensusTime().getEpochSecond() + THREE_MONTHS_IN_SECONDS`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 5]
```diff
@@ -132,17 +163,100 @@ private HederaAccountCustomizer asCustomizer(CryptoCreateTransactionBody op) {
...
-                new HederaAccountCustomizer()
-                        .key(key)
-                        .memo(op.getMemo())
-                        .expiry(expiry)
-                        .autoRenewPeriod(autoRenewPeriod)
-                        .isReceiverSigRequired(op.getReceiverSigRequired())
-                        .maxAutomaticAssociations(op.getMaxAutomaticTokenAssociations())
-                        .isDeclinedReward(op.getDeclineReward());
+        var customizer = new HederaAccountCustomizer();
+        if (!op.getAlias().isEmpty() && !op.hasKey()) {
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:167]: nit: Extract this into a `boolean validAlias = !op.getAlias().isEmpty()` for easy reading in all the scenarios below
```suggestion
        // alias with no key case
        if (validAlias && !op.hasKey()) {
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 6]
```diff
@@ -132,17 +163,100 @@ private HederaAccountCustomizer asCustomizer(CryptoCreateTransactionBody op) {
...
+                        .memo(AUTO_MEMO)
+                        .autoRenewPeriod(THREE_MONTHS_IN_SECONDS)
+                        .expiry(txnCtx.consensusTime().getEpochSecond() + THREE_MONTHS_IN_SECONDS)
+                        .isReceiverSigRequired(false)
+                        .isSmartContract(false)
+                        .alias(op.getAlias());
+            } else {
+                final var keyFromAlias = asPrimitiveKeyUnchecked(op.getAlias());
+                if (!keyFromAlias.getECDSASecp256K1().isEmpty()) {
+                    final var recoveredEvmAddress =
```
Comment (Nana-EC) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:184]: Move the `recoveredEvmAddress` calculation and check into one method as the logic is duped at least 2 more times

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Logic, seems good.
A little hard to follow so recommended readability improvements.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 7]
```diff
@@ -132,17 +166,75 @@ private HederaAccountCustomizer asCustomizer(CryptoCreateTransactionBody op) {
...
+
+                final JKey jKeyFromAlias = asFcKeyUnchecked(keyFromAlias);
+                customizer
+                        .key(jKeyFromAlias)
+                        .memo(AUTO_MEMO)
+                        .autoRenewPeriod(THREE_MONTHS_IN_SECONDS)
+                        .expiry(txnCtx.consensusTime().getEpochSecond() + THREE_MONTHS_IN_SECONDS)
+                        .isReceiverSigRequired(false)
+                        .isSmartContract(false)
+                        .alias(op.getAlias());
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:199]: You can try to remove the code duplication between 179-184 and 194-199.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 8]
```diff
@@ -169,19 +228,87 @@ public ResponseCodeEnum validate(TransactionBody cryptoCreateTxn) {
...
+                return INVALID_ADMIN_KEY;
+            }
+
+            if (!op.getKey().getECDSASecp256K1().isEmpty()) {
+                final var recoveredEvmAddress =
+                        recoverAddressFromPubKey(op.getKey().getECDSASecp256K1().toByteArray());
+                if (!Arrays.equals(recoveredEvmAddress, op.getAlias().toByteArray())) {
+                    return INVALID_ALIAS_KEY;
+                }
+            }
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:309]: Lines 303-309 could be moved to a private method - `validateECDSAKey()`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 9]
```diff
@@ -169,19 +228,87 @@ public ResponseCodeEnum validate(TransactionBody cryptoCreateTxn) {
...
         }
-        if (!fcKey.isValid()) {
-            return INVALID_ADMIN_KEY;
+
+        if (!emptyAlias && op.getAlias().size() == EVM_ADDRESS_SIZE) {
+            var isAliasUsedCheck = isAliasUsedCheck(op.getAlias());
+
+            if (isAliasUsedCheck != OK) {
+                return isAliasUsedCheck;
+            }
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:297]: Lines 293-297 could be moved to a private method - `checkAliasUsed()`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 10]
```diff
@@ -169,19 +228,87 @@ public ResponseCodeEnum validate(TransactionBody cryptoCreateTxn) {
...
+                var evmAddressValidity =
+                        tryToRecoverEVMAddressAndCheckValidity(
+                                asPrimitiveKeyUnchecked(op.getAlias())
+                                        .getECDSASecp256K1()
+                                        .toByteArray());
+
+                if (evmAddressValidity != OK) {
+                    return evmAddressValidity;
+                }
+            }
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:289]: Lines 279-289 could be moved to a private method - checkAliasConversionToECDSAKey()

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 11]
```diff
@@ -169,19 +228,87 @@ public ResponseCodeEnum validate(TransactionBody cryptoCreateTxn) {
...
+
+            if (!op.getKey().getECDSASecp256K1().isEmpty()) {
+                var evmAddressValidity =
+                        tryToRecoverEVMAddressAndCheckValidity(
+                                op.getKey().getECDSASecp256K1().toByteArray());
+
+                if (evmAddressValidity != OK) {
+                    return evmAddressValidity;
+                }
+            }
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:254]: Lines 246-256 could be moved to a private method - `validateEVMAddressRecoverFromECDSAKey()` and it could be also used to substitute lines 269-277.

**Review by Nana-EC — APPROVED** [review]
LG

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 12]
```diff
@@ -119,6 +132,21 @@ public void doStateTransition() {
 
             txnCtx.setCreated(created);
             txnCtx.setStatus(SUCCESS);
+            if (!op.getAlias().isEmpty()) {
+                aliasManager.link(op.getAlias(), EntityNum.fromAccountId(created));
+                if (op.getAlias().size() > EVM_ADDRESS_SIZE) {
+                    final var key = asPrimitiveKeyUnchecked(op.getAlias());
+                    JKey jKey = asFcKeyUnchecked(key);
```
Comment (IvanKavaldzhiev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:139]: jKey could be made `final`

File: test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoCreate.java [hunk 13]
```diff
@@ -83,6 +84,7 @@ public class HapiCryptoCreate extends HapiTxnOp<HapiCryptoCreate> {
     private Optional<String> stakedAccountId = Optional.empty();
     private Optional<Long> stakedNodeId = Optional.empty();
     private boolean isDeclinedReward = false;
+    private Optional<ByteString> withAlias = Optional.empty();
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/transactions/crypto/HapiCryptoCreate.java:87]: The field is no longer boolean. It can be renamed to `alias`.

File: test-clients/src/main/resource/bootstrap.properties [hunk 14]
```diff
@@ -41,6 +41,8 @@ staking.rewardHistory.numStoredPeriods=3
 # Dynamic properties
 accounts.maxNumber=5_000_000
 autoCreation.enabled=true
+lazyCreation.enabled=false
+cryptoCreateWithAlias.enabled=false
```
Comment (IvanKavaldzhiev) [inline test-clients/src/main/resource/bootstrap.properties:45]: Why are the flags turned off?

**Review by IvanKavaldzhiev — APPROVED** [review]
LGTM

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java [hunk 15]
```diff
@@ -119,6 +132,26 @@ public void doStateTransition() {
 
             txnCtx.setCreated(created);
             txnCtx.setStatus(SUCCESS);
+            if (dynamicProperties.isCryptoCreateWithAliasEnabled()
+                    && dynamicProperties.isLazyCreationEnabled()) {
```
Comment (dimitar-dinev) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/services/txns/crypto/CryptoCreateTransitionLogic.java:136]: Won't we miss linking the alias with the new account in the case where we have `cryptoCreateWithAlias` true, `lazy create` = false, and we submit `CryptoCreate` with public key alias?

**Review by dimitar-dinev — COMMENTED** [review]
Just some refactoring suggestions, everything else looks good

**Review by dimitar-dinev — APPROVED** [review]

**Review by qnswirlds — APPROVED** [review]
Reviewed and approved for the changes in gradle files.
